### PR TITLE
JS - NodeJS CWE-444 InsecureHTTPParser

### PR DESCRIPTION
**Pull request by natejohnson05**
In February 2020, NodeJS released a patch to several CVEs related to the way they parsed HTTP requests. The details for that patch exist [here](https://nodejs.org/en/blog/vulnerability/february-2020-security-releases). 

As part of that patch, they introduced a new CLI flag `--insecure-http-parser` and a new option for http and https servers and requests `insecureHTTPParser=true` that allows strict parsing to be disabled. Disabling strict parsing however, exposes these vulnerabilities. It is recommended by NodeJS to not disable strict parsing.

This query checks for all static uses of `insecureHTTPParser` set to true, to warn of the potential vulnerability.

### CODE REVIEW

**Comment by natejohnson05** [comment]
I will add a change note in the morning.

**Comment by erik-krogh** [comment]
Can you move it to [the experimental folder](https://github.com/github/codeql/tree/main/javascript/ql/src/experimental/Security)?

File: javascript/ql/src/Security/CWE-444/InsecureHttpParser.ql [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+/** Gets an options object for an http or https connection. */
+DataFlow::ObjectLiteralNode nodeOptions() { result.flowsTo(nodeInvocation().getAnArgument()) }
+
+from DataFlow::PropWrite disable
+where
+  exists(DataFlow::SourceNode env |
+    env = NodeJSLib::process().getAPropertyRead("env") and
+    disable = env.getAPropertyWrite("NODE_OPTIONS") and
+    disable.getRhs().getStringValue().regexpMatch(".*--insecure-http-parser.*")
```
[the remark this answers is not in the document]
Comment (natejohnson05) [reply javascript/ql/src/Security/CWE-444/InsecureHttpParser.ql]: Fixed this :)

File: javascript/ql/src/experimental/Security/CWE-444/InsecureHttpParser.ql [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+/** Gets an options object for an http or https connection. */
+DataFlow::ObjectLiteralNode nodeOptions() { result.flowsTo(nodeInvocation().getAnArgument()) }
+
+from DataFlow::PropWrite disable
+where
+  exists(DataFlow::SourceNode env |
+    env = NodeJSLib::process().getAPropertyRead("env") and
+    disable = env.getAPropertyWrite("NODE_OPTIONS") and
+    disable.getRhs().getStringValue().matches("%--insecure-http-parser%")
```
Comment (erik-krogh) [inline javascript/ql/src/experimental/Security/CWE-444/InsecureHttpParser.ql:30]: Here you are requiring that `disable.getRhs()` should be a certain string value.  
But in the last line of the `where` you require it to be the boolean value `true`.  
Those two can't be true at the same time.

Comment (natejohnson05) [reply javascript/ql/src/experimental/Security/CWE-444/InsecureHttpParser.ql:30]: My understanding was that the `and` took precedence and thus the flow was treated like this: `where a1 Or (a2 And b)
`. I originally had parenthesis to guarantee this was the case, but the formatter removed them. It proved to produce the desired result in my testing as well. Let me know if I am misunderstanding what is going on.

Comment (natejohnson05) [reply javascript/ql/src/experimental/Security/CWE-444/InsecureHttpParser.ql:30]: The parameter can exist as a server option `insecureHTTPParser=true` or in the `NODE_OPTIONS` as a cli parameter `NODE_OPTIONS="--insecure-http-parser"`. So I have to do both boolean and string comparison.

Comment (erik-krogh) [reply javascript/ql/src/experimental/Security/CWE-444/InsecureHttpParser.ql:30]: Ahhh, right. I didn't notice how the parentheses are grouped.  
It's good 👍

**Review by erik-krogh — APPROVED** [review]
LGTM 👍  

Thanks.

**Pull request merged**